docs/rfcs/017-pageserver-op-atomicity.md

```diff
@@ -0,0 +1,153 @@
+6. Rename local directory in place
+
+If we crash before step 5, S3 may have a timeline metadata file and some
+layer files, without corresponding entry in tenant.json file. That's OK.
+Whenever we see that, we can delete the leftover timeline files.
+
+If we want to make that less scary, we could update a tenant.json file in S3
+twice. First, add the new timeline ID to the file with a flag indicating
+that it's being created. Do that before uploading anything else to S3. And
+then in step 5, update tenant.json to indicate that the creation is complete.
+
+## Branch timeline
+
+Create a new timeline with an existing timeline as parent
+
+1. create timeline directory locally, with .temp extension
+2. create metadata file in the local directory
+3. upload metadata file to S3
+4. update tenant.json file in S3
+5. Rename local directory in place
+
+Like with Create timeline, if we crash between steps 3 and 4, we will
+leave behind a timeline metadata file with no corresponding entry in
+tenant.json.  That's harmless.
+
+## Delete timeline
+
+1. rename local timeline directory to have .temp extension
+2. Update tenant.json file in S3
+3. delete index file from S3
+4. delete layer files from S3
+5. delete local directory
+
+Like with creation, if this is interrupted, we will leave behind
+timeline files in S3 with no corresponding entry in tenant.json. If we
+want to make that less scary, we can update tenant.json in step 2 with
+a tombstone flag for the timeline we're removing, instead of removing
+the entry for it outright.
+
+## Delete tenant
+
+1. rename local tenant directory to have .temp extension
+2. delete tenant.json file in S3
+3. delete all timeline index files from S3
+4. delete all layer files from S3
+5. delete local directory
+
+Like with timeline creation, this can leave behind files with no corresponding
+tenant.json file. We can make it less scary by adding tombstones.
+
+## Attach tenant
+
+1. create local tenant directory with .temp extension
+2. Download tenant.json file
+3. download index files for every timeline
+4. download all layer files (in the future, skip this and download them on demand)
+5. rename local tenant directory in place
+
+## Detach tenant
+
+1. rename local tenant directory to have .temp extension
+2. delete local directory
+
+
+## Load tenant
+
+This happens automatically at pageserver startup, for every tenant that is found
+in the tenants-directory. I.e. for every tenant that was attached to the pageserver
+before the crash or shutdown.
+
+1. download tenant.json file
+2. for every timeline that's in remote tenant.json:
+   1. download remote index file
+   2. download all layer files that are missing locally (skip in future, and download on-demand)
+   3. schedule upload of all files present locally, but missing remotely
+   4. schedule index file upload
+3. delete all locally present timeline directories that's not in tenant.json
+
+
+On startup, delete everything with the .temp extension
```

Comment:
There could be a situation with a few `Attach tenant` operations were cancelled on step `1.` and we restart the pageserver.

If I understand right, current approach would discard such attaches, while what we can actually do is to double check every `.temp` tenant directory for index presence and continue attaching, if such index was found.

Reply:
Yeah, we could do that. It would make sense if the Attach is asynchronous, i.e. if the API call returns to the caller before the attach has finished. The guarantee would be that if the API call has returned Success to the caller, then the pageserver promises to eventually finish the attach operation, even if it's restarted.